docs/src/lucet-runtime/killswitch.md

```diff
@@ -0,0 +1,341 @@
+* `execution_domain`, a `Domain` that describes the kind of execution that is
+  currently happening in the `Instance`. This is kept in a `Mutex` since in
+  many cases it will need to be accessed either by a `KillSwitch` or
+  `KillState`, and for the duration either are considering the domain, it must
+  block other users.
+* `terminable`, an `AtomicBool` that indicates if the `Instance` may stop
+  executing.
+
+The astute observer may notice that `Lucet` contains both an instance `Domain`
+and an instance `State`. These discuss different aspects of the instance's
+lifecycle: `Domain` describes what the instance is doing right now, where
+`State` describes the state of the instance as visible outside
+`lucet_runtime`. This is fundamentally the reason why `State` does not and
+cannot describe a "making a hostcall" variant - this is not something
+`lucet_runtime` wants to expose, and at the moment we don't intend to make any
+commitments about being able to query instance-internal state like this.
+Additionally, this is why `Domain` only expresses that an instance has
+_stopped_ running, not if it successfully or unsuccessfully exited. TODO:
+there's even an argument that `Terminated` vs `Cancelled` is a stronger
+distinction than `Domain` needs to make!
+
+### Termination Mechanisms
+
+At a high level,`KillSwitch` picks one of several mechanisms to terminate an
+instance.  The choice of which mechanism depends on what kind of code is
+running in the guest, indicated by the `instance::execution::Domain` enum.
+
+`Domain::Pending` is the easiest domain to stop execution in: we simply update
+the execution domain to `Cancelled` so when the instance is run we know that we
+in fact should not run it.
+
+`Domain::Guest` is likely to be the next most-immediate termination form, where
+`lucet_runtime` will send a thread-directed `SIGARLM` to the thread running the
+Lucet instance that currently is in guest code.
+
+`Domain::Hostcall` results in the mechanism with least guarantees:
+`lucet_runtime` can't know if it's safe to signal in arbitrary host-provided
+code[1]. Instead, we set the execution domain to `Domain::Terminated` and wait
+for the host code to complete, at which point `lucet_runtime` will exit the
+guest.
+
+#### Guest Signalling
+
+There are two other pieces of information attached to `KillState` that support
+the specific case where we need to send a `SIGALRM`: the thread ID we need to
+signal, and a [`Condvar`][condvar] we can wait on to know when the instance has
+been stopped. The thread ID is necessary because we don't record _where_ the
+instance is running anywhere else, and we keep it here because, so far,
+`KillState` is the only place we actually need to care. Meanwhile, the
+`Condvar` allows us to be smarter than a simple spin loop while waiting for
+signal handling machinery to actually terminate an `Instance` in
+`Domain::Guest`.
+
+### Lifecycle
+
+Having described both `KillState` and the termination mechanisms it helps
```

Comment:
I like the description of this lifecycle.  There's something I feel unsure about that feels a little fundamental and I'm sorry I'm not getting it.  I *think* that every `Instance` has a lifecycle and that every `Instance` has a KillState that tracks that lifecycle in the way that you've described.  If so, I'd love a brief intro sentence to make this more concrete.

Reply:
Ah, you're right that this should be explicitly written here. There's a unit smaller than `Instance` that a `KillState` is associated with, and I don't have a great word for it, which is probably why I skirted around it. The lifecycle and `KillState` described are for one _execution_ of an `Instance` function (one call to `Instance::run`). `KillState` lives on `Instance` because only one execution can happen at a time. A single `Instance` could be terminated dozens of times, because the thing being terminated is the function being run _in_ the `Instance` being terminated.